File: lib/engine/season_progression.dart
class SeasonFixture {
  final int id;
  final int homeTeamId;
  final int awayTeamId;
  final String homeTeamName;
  final String awayTeamName;
  final int? homeGoals;
  final int? awayGoals;
  final String date;
  final String status;

  const SeasonFixture({
    required this.id,
    required this.homeTeamId,
    required this.awayTeamId,
    required this.homeTeamName,
    required this.awayTeamName,
    required this.homeGoals,
    required this.awayGoals,
    required this.date,
    required this.status,
  });

  SeasonFixture copyWith({
    int? id,
    int? homeTeamId,
    int? awayTeamId,
    String? homeTeamName,
    String? awayTeamName,
    int? homeGoals,
    int? awayGoals,
    String? date,
    String? status,
  }) {
    return SeasonFixture(
      id: id ?? this.id,
      homeTeamId: homeTeamId ?? this.homeTeamId,
      awayTeamId: awayTeamId ?? this.awayTeamId,
      homeTeamName: homeTeamName ?? this.homeTeamName,
      awayTeamName: awayTeamName ?? this.awayTeamName,
      homeGoals: homeGoals ?? this.homeGoals,
      awayGoals: awayGoals ?? this.awayGoals,
      date: date ?? this.date,
      status: status ?? this.status,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'homeTeamId': homeTeamId,
      'awayTeamId': awayTeamId,
      'homeTeamName': homeTeamName,
      'awayTeamName': awayTeamName,
      'homeGoals': homeGoals,
      'awayGoals': awayGoals,
      'date': date,
      'status': status,
    };
  }

  factory SeasonFixture.fromJson(Map<String, dynamic> json) {
    return SeasonFixture(
      id: json['id'] as int,
      homeTeamId: json['homeTeamId'] as int,
      awayTeamId: json['awayTeamId'] as int,
      homeTeamName: json['homeTeamName'] as String,
      awayTeamName: json['awayTeamName'] as String,
      homeGoals: json['homeGoals'] as int?,
      awayGoals: json['awayGoals'] as int?,
      date: json['date'] as String,
      status: json['status'] as String,
    );
  }
}

class SeasonWeek {
  final int week;
  final List<SeasonFixture> fixtures;
  final bool completed;

  SeasonWeek({
    required this.week,
    required List<SeasonFixture> fixtures,
    required this.completed,
  }) : fixtures = List.unmodifiable(fixtures);

  SeasonWeek copyWith({
    int? week,
    List<SeasonFixture>? fixtures,
    bool? completed,
  }) {
    return SeasonWeek(
      week: week ?? this.week,
      fixtures: fixtures ?? this.fixtures,
      completed: completed ?? this.completed,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'week': week,
      'fixtures': fixtures.map((fixture) => fixture.toJson()).toList(),
      'completed': completed,
    };
  }

  factory SeasonWeek.fromJson(Map<String, dynamic> json) {
    return SeasonWeek(
      week: json['week'] as int,
      fixtures: (json['fixtures'] as List<dynamic>)
          .map((item) => SeasonFixture.fromJson(item as Map<String, dynamic>))
          .toList(),
      completed: json['completed'] as bool,
    );
  }
}

class LeagueStanding {
  final int position;
  final int clubId;
  final String clubName;
  final int played;
  final int wins;
  final int draws;
  final int losses;
  final int goalsFor;
  final int goalsAgainst;
  final int points;

  const LeagueStanding({
    required this.position,
    required this.clubId,
    required this.clubName,
    required this.played,
    required this.wins,
    required this.draws,
    required this.losses,
    required this.goalsFor,
    required this.goalsAgainst,
    required this.points,
  });

  LeagueStanding copyWith({
    int? position,
    int? clubId,
    String? clubName,
    int? played,
    int? wins,
    int? draws,
    int? losses,
    int? goalsFor,
    int? goalsAgainst,
    int? points,
  }) {
    return LeagueStanding(
      position: position ?? this.position,
      clubId: clubId ?? this.clubId,
      clubName: clubName ?? this.clubName,
      played: played ?? this.played,
      wins: wins ?? this.wins,
      draws: draws ?? this.draws,
      losses: losses ?? this.losses,
      goalsFor: goalsFor ?? this.goalsFor,
      goalsAgainst: goalsAgainst ?? this.goalsAgainst,
      points: points ?? this.points,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'position': position,
      'clubId': clubId,
      'clubName': clubName,
      'played': played,
      'wins': wins,
      'draws': draws,
      'losses': losses,
      'goalsFor': goalsFor,
      'goalsAgainst': goalsAgainst,
      'points': points,
    };
  }

  factory LeagueStanding.fromJson(Map<String, dynamic> json) {
    return LeagueStanding(
      position: json['position'] as int,
      clubId: json['clubId'] as int,
      clubName: json['clubName'] as String,
      played: json['played'] as int,
      wins: json['wins'] as int,
      draws: json['draws'] as int,
      losses: json['losses'] as int,
      goalsFor: json['goalsFor'] as int,
      goalsAgainst: json['goalsAgainst'] as int,
      points: json['points'] as int,
    );
  }
}

class Season {
  final int year;
  final int currentWeek;
  final int totalWeeks;
  final List<LeagueStanding> standings;
  final List<SeasonWeek> weeks;

  Season({
    required this.year,
    required this.currentWeek,
    required this.totalWeeks,
    required List<LeagueStanding> standings,
    required List<SeasonWeek> weeks,
  })  : standings = List.unmodifiable(standings),
        weeks = List.unmodifiable(weeks);

  Season copyWith({
    int? year,
    int? currentWeek,
    int? totalWeeks,
    List<LeagueStanding>? standings,
    List<SeasonWeek>? weeks,
  }) {
    return Season(
      year: year ?? this.year,
      currentWeek: currentWeek ?? this.currentWeek,
      totalWeeks: totalWeeks ?? this.totalWeeks,
      standings: standings ?? this.standings,
      weeks: weeks ?? this.weeks,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'year': year,
      'currentWeek': currentWeek,
      'totalWeeks': totalWeeks,
      'standings': standings.map((standing) => standing.toJson()).toList(),
      'weeks': weeks.map((week) => week.toJson()).toList(),
    };
  }

  factory Season.fromJson(Map<String, dynamic> json) {
    return Season(
      year: json['year'] as int,
      currentWeek: json['currentWeek'] as int,
      totalWeeks: json['totalWeeks'] as int,
      standings: (json['standings'] as List<dynamic>)
          .map((item) => LeagueStanding.fromJson(item as Map<String, dynamic>))
          .toList(),
      weeks: (json['weeks'] as List<dynamic>)
          .map((item) => SeasonWeek.fromJson(item as Map<String, dynamic>))
          .toList(),
    );
  }
}

File: archive/typescript-backup/lib/season-progression.ts
export interface SeasonWeek {
  week: number;
  fixtures: Fixture[];
  completed: boolean;
}

export interface Fixture {
  id: number;
  homeTeamId: number;
  awayTeamId: number;
  homeTeamName: string;
  awayTeamName: string;
  homeGoals?: number;
  awayGoals?: number;
  date: string;
  status: "upcoming" | "completed" | "live";
}

export interface LeagueStanding {
  position: number;
  clubId: number;
  clubName: string;
  played: number;
  wins: number;
  draws: number;
  losses: number;
  goalsFor: number;
  goalsAgainst: number;
  points: number;
}

export interface Season {
  year: number;
  currentWeek: number;
  totalWeeks: number;
  standings: LeagueStanding[];
  weeks: SeasonWeek[];
}

export function generateFixtures(clubCount: number): Fixture[] {
  const fixtures: Fixture[] = [];
  let fixtureId = 1;

  const clubs = Array.from({ length: clubCount }, (_, i) => ({
    id: i + 1,
    name: `Team ${i + 1}`,
  }));

  for (let i = 0; i < clubs.length; i++) {
    for (let j = i + 1; j < clubs.length; j++) {
      const homeClub = clubs[i];
      const awayClub = clubs[j];

      fixtures.push({
        id: fixtureId++,
        homeTeamId: homeClub.id,
        awayTeamId: awayClub.id,
        homeTeamName: homeClub.name,
        awayTeamName: awayClub.name,
        date: new Date().toISOString().split("T")[0],
        status: "upcoming",
      });

      fixtures.push({
        id: fixtureId++,
        homeTeamId: awayClub.id,
        awayTeamId: homeClub.id,
        homeTeamName: awayClub.name,
        awayTeamName: homeClub.name,
        date: new Date().toISOString().split("T")[0],
        status: "upcoming",
      });
    }
  }

  return fixtures;
}

export function organizeFixturesByWeek(fixtures: Fixture[], weeksCount: number): SeasonWeek[] {
  const weeks: SeasonWeek[] = [];
  const fixturesPerWeek = Math.ceil(fixtures.length / weeksCount);

  for (let week = 1; week <= weeksCount; week++) {
    const startIndex = (week - 1) * fixturesPerWeek;
    const endIndex = Math.min(week * fixturesPerWeek, fixtures.length);
    const weekFixtures = fixtures.slice(startIndex, endIndex);

    weeks.push({
      week,
      fixtures: weekFixtures,
      completed: false,
    });
  }

  return weeks;
}

export function updateStandings(
  standings: LeagueStanding[],
  homeTeamId: number,
  awayTeamId: number,
  homeGoals: number,
  awayGoals: number
): LeagueStanding[] {
  const newStandings = standings.map((s) => ({ ...s }));

  const homeTeam = newStandings.find((s) => s.clubId === homeTeamId);
  const awayTeam = newStandings.find((s) => s.clubId === awayTeamId);

  if (!homeTeam || !awayTeam) return standings;

  homeTeam.played += 1;
  homeTeam.goalsFor += homeGoals;
  homeTeam.goalsAgainst += awayGoals;

  awayTeam.played += 1;
  awayTeam.goalsFor += awayGoals;
  awayTeam.goalsAgainst += homeGoals;

  if (homeGoals > awayGoals) {
    homeTeam.wins += 1;
    homeTeam.points += 3;
    awayTeam.losses += 1;
  } else if (homeGoals < awayGoals) {
    awayTeam.wins += 1;
    awayTeam.points += 3;
    homeTeam.losses += 1;
  } else {
    homeTeam.draws += 1;
    homeTeam.points += 1;
    awayTeam.draws += 1;
    awayTeam.points += 1;
  }

  newStandings.sort((a, b) => {
    if (b.points !== a.points) return b.points - a.points;
    return b.goalsFor - b.goalsAgainst - (a.goalsFor - a.goalsAgainst);
  });

  return newStandings.map((s, i) => ({
    ...s,
    position: i + 1,
  }));
}

export function advanceWeek(season: Season): Season {
  if (season.currentWeek >= season.totalWeeks) {
    return season;
  }

  const newSeason = { ...season };
  newSeason.currentWeek += 1;
  newSeason.weeks[season.currentWeek - 1].completed = true;

  return newSeason;
}

export function initializeSeason(year: number, clubCount: number): Season {
  const fixtures = generateFixtures(clubCount);
  const weeks = organizeFixturesByWeek(fixtures, 38);

  const standings: LeagueStanding[] = Array.from({ length: clubCount }, (_, i) => ({
    position: i + 1,
    clubId: i + 1,
    clubName: `Team ${i + 1}`,
    played: 0,
    wins: 0,
    draws: 0,
    losses: 0,
    goalsFor: 0,
    goalsAgainst: 0,
    points: 0,
  }));

  return {
    year,
    currentWeek: 1,
    totalWeeks: 38,
    standings,
    weeks,
  };
}
